### DISCUSSION TITLE
Code versioning

### DISCUSSION THREAD

**Discussion by tuscland** [probabl-ai/skore#573] [Ideas]
Based on: #120

We would like to provide a more general way for users to version their code so that metrics, plots, or anything you save in mandr can be traced back to the code that was responsible for storing it. The other way round could be possible as well, so you can see all the versions of the thing that was stored at a specific line of code.

**Comment by augustebaum** [comment]
A first step could be to maintain a Git repository on behalf of the user, where we do a commit every time a Mandr changes

**Comment by tuscland** [comment]
Are there python libraries we can embed to perform Git operations?

**Comment by rouk1** [comment]
- https://github.com/libgit2/pygit2 which is GPLv2
- https://github.com/gitpython-developers/GitPython which is cool but in maintenance mode and requires user to have git installed
- https://github.com/jelmer/dulwich which is a pure python implementation (and has the best name ever)

**Comment by tuscland** [comment]
It could be an opt-in process:
- The user opts to let Mandr manage code versioning, and optionally the name of a branch where to commit.
- If a repo does not exist, Mandr initializes one.
- Mandr creates a dedicated branch.
